src/fluence/jgf/types.go

```diff
@@ -0,0 +1,62 @@
+/*
+Copyright © 2021 IBM Corporation
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jgf
+
+type node struct {
+	Id       string       `json:"id"`
+	Label    string       `json:"label,omitempty"`
+	Metadata nodeMetadata `json:"metadata,omitempty"`
+}
+
+type edge struct {
+	Source   string       `json:"source"`
+	Relation string       `json:"relation,omitempty"`
+	Target   string       `json:"target"`
+	Directed bool         `json:"directed,omitempty"`
+	Metadata edgeMetadata `json:"metadata"`
+}
+
+type edgeMetadata struct {
+	Name map[string]string `json:"name,omitempty"`
+}
+
+type nodeMetadata struct {
+	Type       string            `json:"type"`
+	Basename   string            `json:"basename"`
+	Name       string            `json:"name"`
+	Id         int               `json:"id"`
+	Uniq_id    int               `json:"uniq_id"`
+	Rank       int               `json:"rank,omitempty"`
```

Comment:
Working on now. One question I have is why for the name, here https://github.com/flux-framework/flux-k8s/blob/9f24f36752e3cced1b1112d93bfa366fb58b3c84/src/fluence/jgf/jgf.go#L116-L158 we use the g.Elements length, but then for others we use an index?

https://github.com/flux-framework/flux-k8s/blob/9f24f36752e3cced1b1112d93bfa366fb58b3c84/src/fluence/jgf/jgf.go#L160-L180

See "Name" in each of the above. I think this would do good to have a comment to explain. My thinking is that maybe there is a subtle distinction between 'socket id in the scope of other sockets" vs "id in the scope of the entire graph" but I think the Name should be the first.

Reply:
The containment paths should also be a path from the cluster root to the named resource. I'm not sure if fluxion needs to use it, if not there is no issue that it's empty. If yes, this might be a source of error. https://github.com/flux-framework/flux-k8s/blob/9f24f36752e3cced1b1112d93bfa366fb58b3c84/src/fluence/jgf/jgf.go#L173-L175

Reply:
okay I'm making a meta issue for us to tackle after this issue here - this entire function I think needs some refactor, but it should be a scoped PR after this one.

Reply:
> My thinking is that maybe there is a subtle distinction between 'socket id in the scope of other sockets" vs "id in the scope of the entire graph" but I think the Name should be the first.

You are correct. See an example here: https://github.com/flux-framework/flux-sched/blob/master/t/data/resource/jgfs/tiny.json.

> The containment paths should also be a path from the cluster root to the named resource. I'm not sure if fluxion needs to use it, if not there is no issue that it's empty. If yes, this might be a source of error.

Fluxion hasn't needed to use it in Fluence yet because paths are used for subsystem checking, and more importantly establishing existence and uniqueness when mutating the resource graph (i.e., elasticity). We should add correct path generation in a follow-up PR.